test/pylib/resource_gather.py

```diff
@@ -0,0 +1,166 @@
+import asyncio
+import getpass
+import logging
+import os
+import platform
+import subprocess
+from abc import ABC
+from asyncio import Task
+from datetime import datetime
+from pathlib import Path
+from typing import TextIO
+
+import psutil
+from attrs import asdict
+
+from test.pylib.db.model import TestMetricRecord, ResourceTimelineRecord
+from test.pylib.db.writer import DATE_TIME_TEMPLATE, SQLiteWriter, RESOURCE_TIMELINE_TABLE, METRICS_TABLE, \
+    DEFAULT_DB_NAME
+
+CGROUP_INITIAL: Path = Path('/sys/fs/cgroup/initial')
+CGROUP_TESTS: Path = CGROUP_INITIAL.parent / 'tests'
+
+
+class ResourceGather(ABC):
+    def __init__(self, test, tmp_dir: str):
+        self.test = test
+        self.db_path = Path(tmp_dir) / DEFAULT_DB_NAME
+        standardized_name = self.test.shortname.replace("/", "_")
+        self.cgroup_path = Path(
+            f"{CGROUP_TESTS}/{self.test.suite.name}.{standardized_name}.{self.test.suite.mode}.{self.test.id}"
+        )
+
+    def make_cgroup(self):
+        pass
+
+    def put_process_to_cgroup(self, pid):
+        pass
+
+    def get_test_metrics(self) -> TestMetricRecord:
+        pass
+
+    def write_metrics_to_db(self, metrics: TestMetricRecord, success: bool = False) -> None:
+        pass
+
+
+class ResourceGatherOff(ResourceGather):
+    pass
+
+
+class ResourceGatherOn(ResourceGather):
+    def __init__(self, test, tmp_dir: str):
+        super().__init__(test, tmp_dir)
+        self.sqlite_writer = SQLiteWriter(self.db_path)
+
+    def make_cgroup(self):
+        os.makedirs(self.cgroup_path, exist_ok=True)
+
+    def put_process_to_cgroup(self, pid):
+        try:
+            with open(self.cgroup_path / 'cgroup.procs', "a") as cgroup:
+                cgroup.write(str(pid))
+        except ProcessLookupError:
+            logging.getLogger().log(
+                level=logging.WARNING,
+                msg=f"Process finished too quick. No metrics will be logged",
+            )
```

Comment:
If the process doesn't exist, would this code trigger a `ProcessLookupError`?
I tried testing this and I got `OSError`

Reply:
Double-checked and got `ProcessLookupError`. If you have `OSError`, please share the details here or in Slack. Maybe you are hitting some edge case and I need to handle it too.